clients/bigquery/storagewrite.go

```diff
@@ -0,0 +1,200 @@
+	"cloud.google.com/go/bigquery/storage/managedwriter/adapt"
+	"github.com/artie-labs/transfer/lib/array"
+	"github.com/artie-labs/transfer/lib/typing"
+	"github.com/artie-labs/transfer/lib/typing/columns"
+	"github.com/artie-labs/transfer/lib/typing/decimal"
+	"github.com/artie-labs/transfer/lib/typing/ext"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+func columnToTableFieldSchema(column columns.Column) (*storagepb.TableFieldSchema, error) {
+	var fieldType storagepb.TableFieldSchema_Type
+	mode := storagepb.TableFieldSchema_NULLABLE
+
+	switch column.KindDetails.Kind {
+	case typing.Boolean.Kind:
+		fieldType = storagepb.TableFieldSchema_BOOL
+	case typing.Integer.Kind:
+		fieldType = storagepb.TableFieldSchema_INT64
+	case typing.Float.Kind:
+		fieldType = storagepb.TableFieldSchema_DOUBLE
+	case typing.EDecimal.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+	case typing.String.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+	case typing.ETime.Kind:
+		switch column.KindDetails.ExtendedTimeDetails.Type {
+		case ext.TimeKindType:
+			fieldType = storagepb.TableFieldSchema_TIME
+		case ext.DateKindType:
+			fieldType = storagepb.TableFieldSchema_DATE
+		case ext.DateTimeKindType:
+			fieldType = storagepb.TableFieldSchema_TIMESTAMP
+		default:
+			return nil, fmt.Errorf("unsupported extended time details type: %q", column.KindDetails.ExtendedTimeDetails.Type)
+		}
+	case typing.Struct.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+	case typing.Array.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+		mode = storagepb.TableFieldSchema_REPEATED
+	default:
+		return nil, fmt.Errorf("unsupported column kind: %q", column.KindDetails.Kind)
+	}
+
+	return &storagepb.TableFieldSchema{
+		Name: column.Name(),
+		Type: fieldType,
+		Mode: mode,
+	}, nil
+}
+
+func columnsToMessageDescriptor(cols []columns.Column) (*protoreflect.MessageDescriptor, error) {
+	fields := make([]*storagepb.TableFieldSchema, len(cols))
+	for i, col := range cols {
+		field, err := columnToTableFieldSchema(col)
+		if err != nil {
+			return nil, err
+		}
+		fields[i] = field
+	}
+	tableSchema := storagepb.TableSchema{Fields: fields}
+
+	descriptor, err := adapt.StorageSchemaToProto2Descriptor(&tableSchema, "root")
+	if err != nil {
+		return nil, fmt.Errorf("failed to build proto descriptor: %w", err)
+	}
+	messageDescriptor, ok := descriptor.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("proto descriptor is not a message descriptor")
+	}
+	return &messageDescriptor, nil
+}
+
+// From https://cloud.google.com/java/docs/reference/google-cloud-bigquerystorage/latest/com.google.cloud.bigquery.storage.v1.CivilTimeEncoder
+// And https://cloud.google.com/pubsub/docs/bigquery#date_time_int
+func encodePacked64TimeMicros(value time.Time) int64 {
+	var result = int64(value.Nanosecond() / 1000)
```

Comment:
`result := int64(...)`

Also can you leave a comment about what you're doing here?

Reply:
Oops I thought I had included a URL that linked to the code I was referencing, let me see if I can find it.